[WIP] Multi-volume snapshots

kadinsayani opened the pull request:
Depends on https://github.com/canonical/lxd/pull/16334.
Based on https://github.com/canonical/lxd/pull/15349.

This pull request introduces support for multi-volume snapshots for instances and their attached volumes, ensuring crash consistency across volumes.

I've gone through each commit and optimized code where possible. There are two `TODO` comments for making loops concurrent; I don't think using concurrency for snapshotting and restoring volumes will help much in those cases since an instance usually does not have many attached volumes and snapshot/restore operations are fairly quick for most storage drivers.

Outstanding todos:
- [ ] expand godoc comments
- [ ] docs
- [ ] lxd-ci tests

kadinsayani commented on test/suites/snapshots.sh:
```diff
@@ -485,3 +485,106 @@ test_snap_fail() {
   lxc delete --force c1
 }
+
+test_snap_multi_volume() {
```
Thanks for the reminder, it's super useful to have `echo`'s to track down point of failures.

kadinsayani commented on test/suites/snapshots.sh:
```diff
@@ -485,3 +485,106 @@ test_snap_fail() {
+  ! lxc restore c1 c1/c1-snap1 --include-attached || false
+
+  # Can restore specific disks.
+  lxc restore c1 c1/c1-snap1 --disks=shared
```
Nice catch. We want to test the presence of `/mnt/shared/snap1` and `snap1` since the root disk is always included.

kadinsayani commented:
With the current implementation, `ParseSnapshotDisks()` is called in the instance snapshot and restore POST handlers. It returns all disk devices sourced by volumes, and these devices are passed to down the call stacks. In the `snapshot/restoreCommon()` functions, if any disks were provided, we call `attachedVolumes()` to get the instance's attached volumes. Then we call `getSharedVolumes()` and error if any volumes are shared with another instance. Both of these functions are quite expensive since they involve db calls and multiple loops to filter volumes. Furthermore, since we want to have an `all-exclusive` mode, I think it makes more sense to only pass disks sourced by exclusive volumes to the snapshot/restore functions. 

I think we can simplify things by adding 2 functions, each with only one db transaction:

```Go
// PlanAttachedVolumesForSnapshot determines the list of custom volumes to snapshot for the provided instance.
// Returns attached volumes to snapshot along with a mapping of volume UUID -> generated snapshot UUID.
func PlanAttachedVolumesForSnapshot(s *state.State, inst Instance, diskVolumes string) (attached []*api.StorageVolume, uuidMap map[string]string, err error)
```

```Go
// ResolveAttachedVolumeSnapshotsForRestore resolves the custom storage volume snapshots that were taken
// alongside sourceSnapshot and are relevant to the provided disk devices. It reads the snapshot's
// "volatile.attached_volumes" (JSON map of volumeUUID -> snapshotUUID) and returns the matching snapshot
// volumes. Fails if a referenced snapshot is missing.
func ResolveAttachedVolumeSnapshotsForRestore(s *state.State, sourceSnapshot Instance, diskVolumes string) ([]*api.StorageVolume, error)
```

These functions can be added to `lxd/instance/instance_utils.go`, which avoids adding multiple helper functions to the already large `lxd/instance/drivers/driver_*.go` files.

I wanted to hear your thoughts @tomponline before I go ahead with this approach. Thanks :)

simondeziel commented on lxc/snapshot.go:
```diff
@@ -44,6 +45,7 @@ running state, including process memory state, TCP connections, ...`))
 	cmd.Flags().BoolVar(&c.flagStateful, "stateful", false, i18n.G("Whether or not to snapshot the instance's running state"))
 	cmd.Flags().BoolVar(&c.flagNoExpiry, "no-expiry", false, i18n.G("Ignore any configured auto-expiry for the instance"))
 	cmd.Flags().BoolVar(&c.flagReuse, "reuse", false, i18n.G("If the snapshot name already exists, delete and create a new one"))
+	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" and "all-exclusive". "root" snapshots the instance's root disk. "all-exclusive" snapshots the instance's root disk and any exclusively attached volumes (non-shared). If not specified, a snapshot is only taken for the root disk. `))
```
```suggestion
	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" (default) and "all-exclusive". "root" only snapshots the instance's root disk. "all-exclusive" snapshots the instance's root disk and any exclusively attached volumes (non-shared).`))
```

simondeziel commented on lxc/restore.go:
```diff
@@ -32,6 +33,7 @@ lxc restore u1 snap0
 
 	cmd.RunE = c.run
 	cmd.Flags().BoolVar(&c.flagStateful, "stateful", false, i18n.G("Whether or not to restore the instance's running state from snapshot (if available)"))
+	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" and "all-exclusive". "root" restores the instance's root disk. "all-exclusive" restores the instance's root disk and any exclusively attached volumes (non-shared) snapshots. If not specified, only the root disk is restored. `))
```
```suggestion
	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" (default) and "all-exclusive". "root" only restores the instance's root disk. "all-exclusive" restores the instance's root disk and any exclusively attached volumes (non-shared) snapshots.`))
```

simondeziel commented on lxc/delete.go:
```diff
@@ -36,6 +38,7 @@ func (c *cmdDelete) command() *cobra.Command {
 	cmd.RunE = c.run
 	cmd.Flags().BoolVarP(&c.flagForce, "force", "f", false, i18n.G("Force the removal of running instances"))
 	cmd.Flags().BoolVarP(&c.flagInteractive, "interactive", "i", false, i18n.G("Require user confirmation"))
+	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" and "all-exclusive". "root" deletes the snapshot for an instance's root disk. "all-exclusive" deletes snapshots for the instance's root disk and any exclusively attached volumes (non-shared). If not specified, only the root disk snapshot is deleted. `))
```
```suggestion
	cmd.Flags().StringVar(&c.flagDiskVolumes, "disk-volumes", "", i18n.G(`Disk volumes mode. Possible values are "root" (default) and "all-exclusive". "root" only deletes the snapshot for an instance's root disk. "all-exclusive" deletes snapshots for the instance's root disk and any exclusively attached volumes (non-shared).`))
```

minaelee reviewed (changes requested):
Just a few unintentional spaces & adding a link target

kadinsayani commented on doc/api-extensions.md:
```diff
@@ -2708,3 +2708,11 @@ This change removes the previous limitation on `ovn` networks that load balancer
 Adds a new `bearer` authentication method and enables authentication to the DevLXD API.
 See {ref}`DevLXD bearer tokens <devlxd-authentication-bearer>`.
+
+## `instance_snapshots_multi_volume`
```
I haven't seen link targets in `doc/api-extension.md`, is this something new we are adding?

tomponline commented on lxd/storage/utils.go:
```diff
@@ -946,29 +946,18 @@ func volumeIsUsedByDevice(vol api.StorageVolume, instanceType instancetype.Type,
-	}
-
-	return false, nil
+	return volumeTypeName == vol.Type && dev["source"] == volName && dev["source.snapshot"] == snapName, nil
```
For the record I'm not a big fan of this style of moving logic into the final return statement because ive found it complicates refactors in the future. Rather than having a if stanza with a return for each stage of the function, the last part is "special" and can lead to mistakes when refactoring. Not a blocker.

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -1280,3 +1280,38 @@ func getSharedVolumes(s *state.State, inst Instance, volumes map[volKey]*db.Stor
+
+// getProjectVolumes gets all custom volumes in the instance's effective
+// project, excluding ISO volumes. Returns a lookup map of storage volumes
+// keyed by [volKey].
```
the wrapped comment line here is odd to me

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -1315,3 +1316,51 @@ func getProjectVolumes(s *state.State, inst Instance) (volumes map[volKey]*db.St
+
+		attached = append(attached, &v.StorageVolume)
+
+		volUUID := v.Config["volatile.uuid"]
```
this volUUID section looks weird, if volUUID is empty then it repeatedly writes a new UUID over the `uuids[""]` key?

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -1315,3 +1316,51 @@ func getProjectVolumes(s *state.State, inst Instance) (volumes map[volKey]*db.St
+	// This is used to resolve snapshots of attached volumes during restore.
+	marshalled, err := json.Marshal(uuids)
+	if err != nil {
+		return nil, err
```
we should wrap error here for more context

tomponline commented on doc/api-extensions.md:
```diff
@@ -2737,3 +2737,11 @@ Adds the field `used_by` to potential storage disk returned by the resources end
+
+Enables the creation of a multi-volume snapshot and the restoration of an instance together with its attached volumes, while ensuring crash consistency across volumes.
+
+Adds a `DiskVolumesMode` field to `POST /1.0/instances/{name}/snapshots` and `PUT /1.0/instances/{name}` to specify which attached volumes to include in a snapshot or restore. "root" mode includes only the instance's root disk, while "all-exclusive" mode includes the root disk and any exclusively attached volumes (non-shared). If not specified, the "root" mode is used by default.
```
Needs to be updated to account for RestoreDiskVolumesMode

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -1364,3 +1364,108 @@ func PlanAttachedVolumesForSnapshot(s *state.State, inst Instance, diskVolumesMo
+	}
+
+	// Get custom storage volumes in the instance's effective project.
+	volumes, err := getProjectVolumes(s, sourceSnapshot)
```
Again, here, really this function should only be retrieving the volumes referenced by the sourceSnapshot instance.

So perhaps this function should be re-factored to parse the supplied instance's devices and load only the referenced volumes.

tomponline commented:
@kadinsayani i made a start on reviewing, but I think we need to get together and discuss again in our 1:1, ill endeavour to read through and come up with a plan by then.

kadinsayani commented on lxd/instance/instance_utils.go:
```diff
@@ -1364,3 +1364,108 @@ func PlanAttachedVolumesForSnapshot(s *state.State, inst Instance, diskVolumesMo
+				continue
+			}
+		}
+
```
errs := d.snapshotRestorePrecheck(sourceSnapshot, diskVolumesMode)

kadinsayani commented on lxd/storage/backend_lxd.go:
```diff
@@ -3694,7 +3708,8 @@ func (b *lxdBackend) CreateInstanceSnapshot(inst instance.Instance, src instance
 	revert.Add(func() { _ = VolumeDBDelete(b, inst.Project().Name, inst.Name(), volType) })
 
 	// Some driver backing stores require that running instances be frozen during snapshot.
-	if b.driver.Info().RunningCopyFreeze && src.IsRunning() && !src.IsFrozen() {
```
Instance driver should already have frozen instance.

kadinsayani commented on lxc/delete.go:
```diff
@@ -64,9 +67,78 @@ func (c *cmdDelete) doDelete(d lxd.InstanceServer, name string) error {
+			return err
+		}
+
+		if c.flagDiskVolumes == api.DiskVolumesModeAllExclusive && len(attachedSnapshotUUIDs) > 0 {
```
Include logic in API, not client side.

tomponline commented on lxd/instance/drivers/driver_lxc.go:
```diff
@@ -2232,7 +2232,7 @@ func (d *lxc) startCommon() (string, []func() error, error) {
 
 	if snapName != "" && expiry != nil {
-		err := d.snapshot(snapName, expiry)
+		err := d.snapshot(snapName, expiry, "")
```
@kadinsayani we should pass in the root volume mode here for explicitness i think.

tomponline commented on lxd/instance/drivers/driver_qemu.go:
```diff
@@ -1449,7 +1449,7 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 
 	if snapName != "" && expiry != nil {
-		err := d.snapshot(snapName, expiry, false)
+		err := d.snapshot(snapName, expiry, false, "")
```
we should pass the root volume mode in here for explicitness.

tomponline commented on lxd/db/storage_volumes.go:
```diff
@@ -171,6 +171,7 @@ type StorageVolumeFilter struct {
 	Project *string
 	Name    *string
 	PoolID  *int64
+	UUIDs   *[]string
```
UUIDs doesnt need to be a pointer as you can have a nil slice and thus it can be made optional without having a pointer.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+			}
+
+			// Attached volume snapshot description.
+			description := "Created alongside " + inst.Type().String() + "/" + inst.Name() + "/" + snapshotName + " in project " + inst.Project().Name
```
lets use a space between instance type and instance name

tomponline replied:
lets not call inst.Project(), inst.Type() and inst.Name() in a loop for efficiency

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
 		return err
 	}
 
+	// Snapshot root disk.
```
Today in `pool.CreateInstanceSnapshot` there's this logic:

```go
	// Some driver backing stores require that running instances be frozen during snapshot.
	if b.driver.Info().RunningCopyFreeze && src.IsRunning() && !src.IsFrozen() {
		// Freeze the processes.
		err = src.Freeze()
		if err != nil {
			return err
		}

		defer func() { _ = src.Unfreeze() }()

		// Attempt to sync the filesystem.
		err = filesystem.SyncFS(src.Path())
		if err != nil {
			l.Warn("Failed to flush writes to instance volume", logger.Ctx{"err": err})
		}
	}
```

Now, `pool.CreateInstanceSnapshot()` is only called from this function today, so I am thinking we should move this logic up into this function so we keep all of the instance freezing and filesystem freezing logic together in `snapshotCommon()`.

This will also reduce multiple checks/calls to `inst.IsRunning()` and `inst.IsFrozen()` when doing multi-volume snapshots.

tomponline replied:
interestingly, `CreateCustomVolumeSnapshot()` has this today too:

```go
	// If the volume is a filesystem and is mounted, attempt to sync the filesystem before taking the snapshot.
	// If RunningCopyFreeze is false for the driver in use, it means the driver syncs the volume on snapshot,
	// so we don't have to do it here.
	if parentVol.ContentType == cluster.StoragePoolVolumeContentTypeNameFS && b.driver.Info().RunningCopyFreeze && filesystem.IsMountPoint(mountPath) {
		err = filesystem.SyncFS(mountPath)
		if err != nil {
			l.Warn("Failed to flush writes to custom volume", logger.Ctx{"err": err})
		}
	}
```

So how about this:

1. We move the instance freezing/unfreezing logic from `pool.CreateInstanceSnapshot` into `snapshotCommon` as I suggested above, but...
2. We keep the `filesystem.SyncFS` calls in `pool.CreateInstanceSnapshot` and `pool.CreateInstanceSnapshot`  respectively so that they can sync the associated mount path of the relevant volume (but only if mounted).

That feels to me to keep the volume specific snapshot logic nicely separated from the instance snapshot logic.

kadinsayani replied:
Thanks for the suggestion! I've made those changes.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+	// Snapshot attached disk volumes (if requested).
+	if diskVolumesMode == api.DiskVolumesModeAllExclusive {
+		// Freeze instance for crash-consistent snapshots.
+		if inst.IsRunning() && !inst.IsFrozen() {
```
This is another reason why I think we need to take the instance freezing logic out of `pool.CreateInstanceSnapshot` because this is not going to cause a consistent snapshot to be taken.

Consider the scenario when `diskVolumesMode == api.DiskVolumesModeAllExclusive`:

1. Call `pool.CreateInstanceSnapshot` - this freezes the instance, snapshots the root volume and unfreezes the instance. So now the instance is running again, writing to disks etc, diverging from snapshot we just took.
2. Then the instance is frozen again and the custom volume snapshots are taken, but they are not consistent with the root volume snapshot taken in 1. because the instance was unfrozen.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+			attachedVolumeUUIDs[volume.Config["volatile.uuid"]] = snapshotUUID.String()
+
+			revert.Add(func() {
+				_ = pool.DeleteCustomVolumeSnapshot(volume.Project, volume.Name+"/"+snapshotName, d.op)
```
@kadinsayani please get into the habit of capturing errors in reverts and logging them as warnings, thanks

kadinsayani replied:
Ack :)

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+		storageCache := storagePools.NewStorageCache(pool) // Create storage cache for pool lookups.
+		attachedVolumeUUIDs := make(map[string]string)
+		for _, volume := range volumes {
+			logger.Debug("Creating attached volume snapshot", logger.Ctx{"volumeProject": volume.Project, "volumePool": volume.Pool, "volumeName": volume.Name})
```
Lets stick with logging conventions of the fields: `pool`, `volume`, `project`.

Also please use d.logger so we get the instance context.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -717,6 +717,51 @@ func (d *common) runHooks(hooks []func() error) error {
+		for _, vol := range volumes {
+			_, _, isSnap := api.GetParentAndSnapshotName(vol.Name)
+			if !isSnap {
+				return fmt.Errorf("Volume %q is not a snapshot", vol.Name)
```
should we log also the volume uuid here so we can understand how it got loaded from the supplied uuids?

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1090,6 +1090,91 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
+	}
+
+	// Get attached volumes.
+	attachedVolumes, err := d.getAttachedVolumes(source)
```
IIUC correctly, `source` is the snapshot being restored right?

But what happens if the instance snapshot's device config references a volume that no longer exists, won't the snapshot restore just fail at this point?

kadinsayani replied:
Correct, `source` is the snapshot being restored. We should instead pass `inst` to `getAttachedVolumes()` so we can check the instance's volumes against `volatile.attached_volumes`.

tomponline replied:
But my question was what if a volume (and its snapshots) were deleted, would this prevent restore?

kadinsayani replied:
Right, `source` is correct.

kadinsayani replied:
I'll make sure the tests cover this.

kadinsayani replied:
Yes, so we will only return an error if a snapshot is missing or a volume is shared.

tomponline replied:
Or we do fail on missing snapshot, and we make the instance snapshot's volatile key editable so the user can remove the offending snapshots?

kadinsayani replied:
Yes, shall we proceed and log warnings if we identify any missing snapshots?

kadinsayani replied:
We commented at the same time.

Editing the volatile key seems like it goes against the current policy of volatile keys being intended for internal use only.

kadinsayani replied:
I think proceeding optimistically and logging warnings along the way would work well.

tomponline replied:
Let's do this for now. Original plan was to fail on missing snapshot and then allow selective restore with the --exclude... flag but that got dropped.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1090,6 +1090,91 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
+				parentName, _, _ := api.GetParentAndSnapshotName(source.Name())
+
+				// Skip this instance.
+				if instanceProject.Name == dbInst.Project && parentName == dbInst.Name {
```
Can we use `instance.IsSameLogicalInstance()` here?

kadinsayani replied:
Yes, we just need to pass the instance to `resolveRestoreSnapshots()` and pass that struct value to `instance.IsSameLogicalInstance()`.

tomponline commented on lxd/instance/drivers/driver_lxc.go:
```diff
@@ -3383,12 +3383,31 @@ func (d *lxc) Restore(sourceContainer instance.Instance, stateful bool, diskVolu
+
+			pool, err := storageCache.GetPool(volume.Pool)
+			if err != nil {
+				return err
```
we need some error wrapping here

tomponline commented on test/suites/snapshots.sh:
Please can you add some tests with a project that doesnt have its own custom volumes too. Ta

tomponline commented on doc/howto/instances_backup.md:
```diff
@@ -55,6 +55,8 @@ Use the following command to create a snapshot of an instance:
 For virtual machines, you can add the `--stateful` flag to capture not only the data included in the instance volume but also the running state of the instance.
 Note that this feature is not fully supported for containers because of CRIU limitations.
+
+By default, instance snapshots include a snapshot of the instance's root disk device only. To include snapshots of attached storage volumes, set the `--disk-volumes` flag to "all-exclusive".
```
mixing terminology here, "device" and "volume".

> instance snapshots include a snapshot of the instance's root disk device only

A "device" is just the config that points to the volume. The reality is that we snapshot both today.

> To include snapshots of attached storage volumes

This would make more sense if we had mentioned the "root volume" above, but as we only mentioned "device", it seems confusing to me.

kadinsayani replied:
I've gone with consistent use of the term "volume".

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1090,6 +1090,91 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
+	for _, v := range attachedVolumes {
+		snapshotUUID, ok := attachedVolumeUUIDs[v.Config["volatile.uuid"]]
+		if !ok {
+			// Skip volumes not included in the source snapshot.
```
Since we're passing `source` to `getAttachedVolumes()` this isn't actually a valid scenario. I'll drop this bit.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -714,6 +716,60 @@ func (d *common) runHooks(hooks []func() error) error {
+func (d *common) getAttachedVolumes(inst instance.Instance) (volumes []*db.StorageVolume, err error) {
+	// Get attached disk volume devices.
+	attachedDiskVolumeDevices := deviceConfig.Devices{}
+	maps.Copy(attachedDiskVolumeDevices, d.expandedDevices.Filter(filters.IsCustomVolumeDisk))
```
why is it copied here rather than iterating over it?

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -822,6 +822,17 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+			return err
+		}
+
+		defer func() { _ = inst.Unfreeze() }()
```
please get into the habit of logging warnings when errors occur in defer/reverts.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -843,6 +846,69 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+		}
+	}
+
 	revert.Add(func() {
```
This reverter needs to be added immediately after successful completion of `pool.CreateInstanceSnapshot` above otherwise in all-exclusive disk mode it opens the possibility of a failure occurring before the revert is defined and leaving the instance snapshot behind.

kadinsayani replied:
Ack, I'll move that bit up.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -717,6 +717,51 @@ func (d *common) runHooks(hooks []func() error) error {
+		for _, vol := range volumes {
+			_, _, isSnap := api.GetParentAndSnapshotName(vol.Name)
+			if !isSnap {
+				return fmt.Errorf("Volume %q with uuid %q is not a snapshot", vol.Name, vol.Config["volatile.uuid"])
```
nice

tomponline commented on shared/util.go:
```diff
@@ -974,6 +974,35 @@ func ParseMetadata(metadata any) (map[string]any, error) {
 
+// ParseJSONConfig extracts and parses a JSON-encoded configuration value from a config map.
+// It returns the parsed configuration as a map[string]string or an error if the key is missing or the JSON is malformed.
+func ParseJSONConfig(key string, configMap map[string]string) (map[string]string, error) {
```
I dont think we need this very generic function in the shared (thus encouraging usage out of LXD) package.

Can we just do:

```go

	var parsed map[string]string
	err := json.Unmarshal([]byte(configMap[key]), &parsed)
	if err != nil {
		return nil, fmt.Errorf("Failed parsing %q: %w", key, err)
	}
```

tomponline commented on lxd/instance/drivers/driver_lxc.go:
```diff
@@ -3383,12 +3383,31 @@ func (d *lxc) Restore(sourceContainer instance.Instance, stateful bool, diskVolu
 		return err
 	}
 
+	if diskVolumesMode == api.DiskVolumesModeAllExclusive {
```
Hrm, so this approach immediately triggered my "symmetry mismatch" detector (for want of a better description). What do I mean?

Well, you have implemented taking the multi-volume snapshot in `snapshotCommon`: 

https://github.com/canonical/lxd/pull/16336/commits/3f6f92ad5108ecaaedb09fef743025832596aa7a

But here, the restoration of the multi volume snapshot has been pushed up (and duplicated) to the specific instance drivers.

So this isn't symmetric in terms of depth of code (instance drivers -> instance common -> storage drivers etc).

So I asked myself: 

- Why is this? 
- What's different about containers and VMs that means we can't do multi volume snapshot restore in `restoreCommon`?


The only difference I can see is the call to:

```go
d.stopForkfile(false)
```

in the lxc driver after `restoreCommon` but before  `pool.RestoreInstanceSnapshot`.

So here's what I think we should do:

1. Move the `d.stopForkfile(false)` call to after `d.logger.Info("Restoring instance", ctxMap)` but before `d.restoreCommon` call.
2. Move the attached volume restore logic into `d.restoreCommon` (including de-duping the call to `pool.RestoreInstanceSnapshot` in each driver).

This achieves de-duplication and is then symmetric with what we do when taking the instance snapshot.

tomponline commented on client/lxd_instances.go:
```diff
@@ -2281,14 +2281,26 @@ func (r *ProtocolLXD) MigrateInstanceSnapshot(instanceName string, name string,
+	_, instanceType, _ := strings.Cut(path, "/")
+	u := api.NewURL().Path(instanceType, instanceName, "snapshots", name).WithQuery("disk-volumes", diskVolumesMode)
+
+	if diskVolumesMode != "" {
```
it can be non-empty as long as its set to "root" right?

kadinsayani replied:
`if diskVolumesMode == api.DiskVolumesModeExclusive`

tomponline commented on lxd/instance/instance_interface.go:
```diff
@@ -103,6 +103,7 @@ type Instance interface {
 	Update(newConfig db.InstanceArgs, userRequested bool) error
 
 	Delete(force bool) error
+	DeleteSnapshot(diskVolumesMode string) error
```
This is good, but please can you update the implementations of `Delete()` to ensure they refuse to operate on snapshot instances, otherwise we're going to end up with 2 ways to delete instance snapshots, and only one of them will be correct.

kadinsayani replied:
Done :)

tomponline commented on test/suites/projects.sh:
```diff
@@ -246,6 +246,33 @@ test_projects_snapshots() {
+  lxc storage volume attach "${pool}" testvol c1 /mnt
+
+  echo "Check snapshot fails with disk volumes mode set to all-exclusive"
+  ! lxc snapshot c1 --disk-volumes=all-exclusive || false
```
can you tweak this to check the specific error message please

tomponline commented on lxd/instance_snapshot.go:
```diff
@@ -328,6 +329,11 @@ func instanceSnapshotsPost(d *Daemon, r *http.Request) response.Response {
 	}
 
+	// When "features.storage.volumes" is disabled, only allow root disk to be snapshotted.
+	if shared.IsFalse(p.Config["features.storage.volumes"]) && req.DiskVolumesMode == api.DiskVolumesModeAllExclusive {
```
Do we need to equivalent of this in restore volume API call?

kadinsayani replied:
Yes, I'll add tests as well.

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -843,6 +846,69 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
 
+	// Snapshot attached disk volumes (if requested).
+	if diskVolumesMode == api.DiskVolumesModeAllExclusive {
+		volumes, err := d.getAttachedVolumes(inst)
```
Good idea, thanks!

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1188,6 +1188,51 @@ func (d *common) resolveRestoreSnapshots(inst instance.Instance, source instance
+		return err
+	}
+
+	pool, err := storagePools.LoadByInstance(d.state, snapInst)
```
Thanks for pointing that function out, it will work well in both snapshot and restore common functions.

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1097,6 +1097,83 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
+// resolveRestoreSnapshots returns a list of snapshot volumes to include in an instance restore.
+// When detectSharing is true, the function will check that the volumes are not shared with other instances.
+func (d *common) resolveRestoreSnapshots(inst instance.Instance, source instance.Instance, detectSharing bool) (restoreSnapshots []*db.StorageVolume, err error) {
+	attachedVolumeUUIDs, err := shared.ParseJSONConfig("volatile.attached_volumes", source.LocalConfig())
```
Ack

kadinsayani commented:
> Thanks @kadinsayani !
> 
> 
> 
> We are very close to merging this now, honestly I went in with a view that I was  going to merge this today and ask for a few follow ups.
> 
> 
> 
> But after going through it I feel there is sufficient changes needed (all small) to hold back another 24 hours to avoid having a confused git history where we land things and then remove them shortly after.

Thanks again for your review! I've addressed all of the comments, but there's still some work to be done with the `Delete*` interface changes to get the tests passing. I'll resume work on this tomorrow and hopefully have the tests passing before our 1:1.

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1103,6 +1103,84 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
 }
 
+// resolveRestoreSnapshots returns a list of snapshot volumes to include in an instance restore.
+// When detectSharing is true, the function will check that the volumes are not shared with other instances.
```
Remove detectSharing param for now.

kadinsayani commented:
@tomponline should be good to go, all tests are passing. Please note that https://github.com/canonical/lxd/pull/16336/commits/5d6254d59916925513b8c97e67ea883279b7ec64 is new. I've extracted a `deleteCommon()` function to de-duplicate shared logic in lxc and qemu `Delete()` implementations.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -825,7 +825,12 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
 
-		defer func() { _ = inst.Unfreeze() }()
+		defer func() {
+			err = inst.Unfreeze()
```
needs `err :=` please

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -703,6 +703,71 @@ func (d *common) rebuildCommon(inst instance.Instance, img *api.Image, op *opera
+		parentName, _, _ := api.GetParentAndSnapshotName(inst.Name())
+
+		// Load the parent.
+		parent, err := instance.LoadByProjectAndName(d.state, d.project.Name, parentName)
```
as this can fail, it would be better to load this before deleting the snapshot.

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -1106,6 +1117,26 @@ func (d *common) restoreCommon(inst instance.Instance, source instance.Instance)
+
+			err = pool.RestoreCustomVolume(volume.Project, volName, snapName, d.op)
+			if err != nil {
+				return false, nil, fmt.Errorf("Failed restoring volume %q snapshot %q: %w", volume.Name, snapName, err)
```
this error should include the pool name as it may be different than the instance's pool name

tomponline commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -850,6 +863,64 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
+			expiry := snap.ExpiryDate() // Attached volume snapshots inherit the expiry date of the instance snapshot.
+			snapshotUUID, err := pool.CreateCustomVolumeSnapshot(volume.Project, volume.Name, snapshotName, description, &expiry, d.op)
+			if err != nil {
+				return err
```
lets wrap this error and be sure to include volume name and pool name as it may be different than the instance's pool name

tomponline commented on client/lxd_instances.go:
```diff
@@ -2295,14 +2295,26 @@ func (r *ProtocolLXD) MigrateInstanceSnapshot(instanceName string, name string,
 		return nil, err
 	}
 
+	_, instanceType, _ := strings.Cut(path, "/")
```
would strings.TrimPrefix be clearer here?

tomponline reviewed (approved):
Thanks @kadinsayani im going to go ahead and merge this as there's only some minor follow ups that need to be addressed. LGTM!

kadinsayani commented on lxd/instance/drivers/driver_common.go:
```diff
@@ -703,6 +703,48 @@ func (d *common) rebuildCommon(inst instance.Instance, img *api.Image, op *opera
+		return fmt.Errorf(`Failed parsing snapshot instance "volatile.attached_volumes": %w`, err)
+	}
+
+	if len(attachedVolumeUUIDs) == 0 {
```
We can get rid of this error scenario since `getAttachedVolumeSnapshots()` performs the same check.
